[JIT] Refactor SchemaCheckMode to check for mutated metadata

ttanpcs opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79991

- Refactored SchemaCheckMode so that arguments and outputs are only unflattened/unwrapped a single time 
- Added cloned_metadata variable inside torch_dispatch which accounts for ops that mutate stride or storage()._cdata
- Fixed aliasing check to compare with any iterable outputs (ie list of tensor output)
- Fixed aliasing check to check ops with an (a -> *) notation argument correctly
- Changed order of mutation and aliasing checks to optimize error message usefulness (generally something incorrectly aliasing implies that something is also incorrectly mutating so an aliasing error gives more information)
- Added tests for resize_, as_strided to ensure that mutated metadata checking works as intended

Commit:
[JIT] Refactor SchemaCheckMode to check for mutated metadata

[ghstack-poisoned]

Commit:
Update on "[JIT] Refactor SchemaCheckMode to check for mutated metadata"


- Refactored SchemaCheckMode so that arguments and outputs are only unflattened/unwrapped a single time 
- Added cloned_metadata variable inside torch_dispatch which accounts for ops that mutate stride or storage()._cdata
- Fixed aliasing check to compare with any iterable outputs (ie list of tensor output)
- Changed order of mutation and aliasing checks to optimize error message usefulness (generally something incorrectly aliasing implies that something is also incorrectly mutating so an aliasing error gives more information)
- Added tests for resize_, as_strided to ensure that mutated metadata checking works as intended

[ghstack-poisoned]

davidberard98 commented on torch/testing/_internal/schema_check_mode.py:
```diff
@@ -72,68 +80,83 @@ def unwrap(e):
+                else:
+                    return (deepcopy(e.stride()), e.storage()._cdata)
+            else:
+                return e
```
nit: it looks like the value will never be used if e is not a tensor... if that's true could we return `None` instead?

goldenxuett replied:
Yup, I had this from an old version where I was running into issues with having None instead, but it seems to work fine now

davidberard98 commented on torch/testing/_internal/schema_check_mode.py:
```diff
@@ -47,11 +54,12 @@ def is_mutable(arg_alias_pairs):
-                if arg.alias_info is not None and bool(len(output_alias_info.after_set & arg.alias_info.after_set)):
-                    return True
+                if arg.alias_info is not None:
+                    if '*' in arg.alias_info.after_set:
```
cc @eellison just to make sure this is the right logic for wildcards

Gamrix replied:
You still need a check that the types are the same. Wildcard aliases do not alias objects of different type classes. See `alias_analysis.h:40`.

eellison replied:
The semantics of `chunk(Tensor(a -> *) self, int chunks, int dim=0) -> Tensor(a)[]` is that the input tensor storage will be contained in the output list

davidberard98 reviewed (approved):
looks good! just wanted to ask @eellison to take a look at the wildcard logic so please wait for his review first

(also thanks for adding the wildcard stuff, I had almost forgotten about it)

Gamrix commented on torch/testing/_internal/schema_check_mode.py:
```diff
@@ -72,68 +80,83 @@ def unwrap(e):
-                unwrap,
-                tree_flatten(out[j])[0])
-            if are_args_aliasing(i_values, j_values) and not bool(len(out_alias_pairs_map[i] & out_alias_pairs_map[j])):
+            if are_args_aliasing(u_out[i], u_out[j]) and not bool(len(out_alias_pairs_map[i] & out_alias_pairs_map[j])):
```
Separate this if statement out into multiple lines, and give 
`bool(len(out_alias_pairs_map[i] & out_alias_pairs_map[j]))` a variable name

Gamrix commented on torch/testing/_internal/schema_check_mode.py:
```diff
@@ -34,11 +35,17 @@ def display_ops(self):
 
         def has_aliased(lhs, rhs):
-            return torch._C._is_alias_of(lhs, rhs) if isinstance(lhs, torch.Tensor) and isinstance(rhs, torch.Tensor) else False
+            return torch._C._is_alias_of(lhs, rhs) if type(lhs) == torch.Tensor and type(rhs) == torch.Tensor else False
```
Why did you change this from an `isinstance()` check to instead check type equality? 

Also, this line is very long and should just be a normal if statement.

goldenxuett replied:
This line was changed from isinstance() to check type equality because it is possible for a subclass of torch.Tensor to pass through the unwrapping due to not having an "elem" instance variable which will then error out on the subsequent mutation check / aliasing check. Changing the checks to type equality ensures this doesn't happen.

Gamrix replied:
Do you have some example code handy for where this would happen? I assume that we want to still check for aliasing information for tensor subclasses. 

(Also, this is good stuff to add as a comment to this function).

goldenxuett replied:
It seems to happen with a "Parameter" object which has no attribute 'elem' when calling the batch norm op. For all other tensor subclasses which have a tensor within them, aliasing information is checked as expected.

Commit:
Update on "[JIT] Refactor SchemaCheckMode to check for mutated metadata"


- Refactored SchemaCheckMode so that arguments and outputs are only unflattened/unwrapped a single time 
- Added cloned_metadata variable inside torch_dispatch which accounts for ops that mutate stride or storage()._cdata
- Fixed aliasing check to compare with any iterable outputs (ie list of tensor output)
- Fixed aliasing check to check ops with an (a -> *) notation argument correctly
- Changed order of mutation and aliasing checks to optimize error message usefulness (generally something incorrectly aliasing implies that something is also incorrectly mutating so an aliasing error gives more information)
- Added tests for resize_, as_strided to ensure that mutated metadata checking works as intended

[ghstack-poisoned]

bdhirsh commented on torch/testing/_internal/schema_check_mode.py:
```diff
@@ -70,70 +85,86 @@ def unwrap(e):
-                return e
+            return e
+
+        def parse_metadata(e):
```
Is the idea here that you want to assert that both the tensor data and metadata are unchanged after performing an op?

Are you able to use the `assertEqual()` method from the testing framework ([here](https://github.com/pytorch/pytorch/blob/09f79e94acce30e2002063322c9fe50d116361f1/torch/testing/_internal/common_utils.py#L2210)) instead? It should already handle checking metadata for you.

Another reason to use the one from the framework is that keeping track of the full set of metadata to check is tough, and can change over time. It looks like today, that method also checks `layout` and `device`: https://github.com/pytorch/pytorch/blob/09f79e94acce30e2002063322c9fe50d116361f1/torch/testing/_comparison.py#L635.

goldenxuett replied:
From what I understand, the issue with assertEqual() is that torch.clone() doesn't copy the strides and storage cdata, so that assertEqual might return true when the strides have in fact changed or vice versa. Thus the parse_metadata function is there specifically to record data that isn't copied correctly in torch.clone().

davidberard98 replied:
I agree with @goldenxuett , I think we need a separate check for these, because `clone()` doesn't copy strides or layout, and so we need to check those properties separately from the `assertEquals`

davidberard98 reviewed (approved):
LGTM! You might need to rebase before it lets you merge it.

goldenxuett commented:
@pytorchbot merge

Commit:
Update on "[JIT] Refactor SchemaCheckMode to check for mutated metadata"


- Refactored SchemaCheckMode so that arguments and outputs are only unflattened/unwrapped a single time 
- Added cloned_metadata variable inside torch_dispatch which accounts for ops that mutate stride or storage()._cdata
- Fixed aliasing check to compare with any iterable outputs (ie list of tensor output)
- Fixed aliasing check to check ops with an (a -> *) notation argument correctly
- Changed order of mutation and aliasing checks to optimize error message usefulness (generally something incorrectly aliasing implies that something is also incorrectly mutating so an aliasing error gives more information)
- Added tests for resize_, as_strided to ensure that mutated metadata checking works as intended

[ghstack-poisoned]

goldenxuett commented:
@pytorchbot merge
